library/core/src/iter/adapters/map_windows.rs

```diff
@@ -0,0 +1,379 @@
+        assert!(N != 0, "array in `Iterator::map_windows` must contain more than 0 elements");
+
+        // Only ZST arrays' length can be so large.
+        if mem::size_of::<I::Item>() == 0 {
+            assert!(
+                N.checked_mul(2).is_some(),
+                "array size of `Iterator::map_windows` is too large"
+            );
+        }
+
+        Self { inner: MapWindowsInner::new(iter), f }
+    }
+}
+
+impl<I: Iterator, const N: usize> MapWindowsInner<I, N> {
+    #[inline]
+    fn new(iter: I) -> Self {
+        Self { iter, initialized: false, buffer: None }
+    }
+
+    #[inline]
+    fn current_window(&self) -> Option<&[I::Item; N]> {
+        self.buffer.as_ref().map(Buffer::as_array_ref)
+    }
+
+    fn next_window(&mut self) -> Option<&[I::Item; N]> {
+        if !self.initialized {
+            // If uninitialized, collect the first `N` items from `self.iter`
+            // to initialize `self.buffer`.
+            self.initialized = true;
+            self.buffer = Buffer::try_from_iter(&mut self.iter).ok();
+        } else if let Some(ref mut buffer) = self.buffer {
+            // Advance iterator. We first call `next` before changing our buffer at
+            // all. This means that if `next` panics, our invariant is upheld and
+            // our `Drop` impl drops the correct elements.
+            if let Some(item) = self.iter.next() {
+                buffer.push(item);
+            } else {
+                // `self.iter.next()` yields `None`, and according to the assumption
+                // that `self.iter` is regarded as fused, drop the buffer to let
+                // `next_window` yields `None` forever.
+                let _ = self.buffer.take();
+            }
+        }
+        self.current_window()
+    }
+
+    fn advance_by(&mut self, n: usize) -> Result<(), usize> {
+        // Helper functions for calculating the offset of steps.
+        fn add_offset(n: usize) -> impl Fn(usize) -> usize + Copy {
+            move |i| i + n
+        }
+
+        if !self.initialized {
+            // If uninitialized, the collected result will be the
+            // `n..n + N`-th items yielded by `self.iter.next()`.
+
+            // We advance the `iter` by `n - 1` steps, requiring `n > 0`.
+            if let Some(steps) = n.checked_sub(1) {
+                self.initialized = true;
+
+                // Notice that the first `N` items are counted as one window,
+                // and thus the first `N` step of the inner iterators is counted
+                // as only one step for the outer iterator.
+                let exclude_first_window = |i: usize| i.saturating_sub(N - 1);
+                self.iter.advance_by(steps).map_err(exclude_first_window)?;
+                self.buffer = Some(
+                    Buffer::try_from_iter(&mut self.iter)
+                        .map_err(add_offset(steps))
+                        .map_err(exclude_first_window)?,
+                );
```

Comment:
Something feels a little off here - I think we would expect that we successfully initialize and (eventually) return a Buffer if it's elements would be [x, x, x, to_be_returned] where x is an element currently advance_by'd before we start accumulating into the buffer. But I think this code would drop those x elements in the self.iter.advance_by, and so we would not match behavior with calling next() n times.

But I'm not quite sure, this feels tricky - I think we want a test at minimum that this behaves as expected, maybe you can point me to an existing one or add one.

Taking a step back, I think the thing that's making me feel weird is that we're only subtracting 1 from `n` to compute steps, not `N` or `N-1`, which would allow us to fill the buffer with the tail, as seems necessary for yielding the `n + 1`th element.

Reply:
> But I think this code would drop those x elements in the self.iter.advance_by, and so we would not match behavior with calling next() n times.

I'm a little confused. Do you mean that whether those `x` elements (in `[x, x, x, to_be_returned]`) are yielded by `self.iter.next()` for `n - 1` times or by `self.iter.advance_by(n - 1)` differs because the former yields and drops those elements but the latter just skips them in place?

Reply:
> But I'm not quite sure, this feels tricky - I think we want a test at minimum that this behaves as expected, maybe you can point me to an existing one or add one.

There is a test of `advance_by`: [link](https://github.com/rust-lang/rust/pull/94667/files#diff-82642196d9d615b4dcb26384c7d7d8f4a66746358409a21ad27c9926a034f9bdR312).

It tests the `advance_by` called at the very beginning before any `.next()` calls, and also after some `.next()` calls.

Reply:
> Taking a step back, I think the thing that's making me feel weird is that we're only subtracting 1 from `n` to compute steps, not `N` or `N-1`, which would allow us to fill the buffer with the tail, as seems necessary for yielding the `n + 1`th element.

Subtracting 1 from `n` is because what we would made for the next window is `n..n + N`th elements (counted from zero), and `Buffer::try_from_iter` should count for one `.next()`, so we need to subtract 1.

Reply:
Oh, I think using `Buffer::try_from_iter(self.iter.by_ref().skip(n - 1))` makes more sense here.

Reply:
Oh, I realize what you have said. You meant that `.next()` should invoke the mapping function `f`, but we would skip that by calling `.advance_by` of the inner iterator. They are not equivalent, right?

I think we'd better remove the overwritten `.advance_by` method and leave it the default implementation.